### PR TITLE
Lowering of subgroups down to vulkan

### PR DESCRIPTION
**Pull request by jbruestle**
Initial reworking to connect all the passes end-to-end.  As of this commit, a 2d convolution lowers into what looks roughly like correct subgroup SPIRV code, and gets all the way to the runtime, at which point we segfault in VulkanRuntime::createComputePipeline.  Lots of rough code and I broke some tests that need fixing, but wanted to put up the current state for today.

To reproduce the failure run:

PLAIDML_VERBOSE=1 PLAIDML_DUMP=/tmp/ir/ PLAIDML_TARGET="intel_gen" bazelisk run -c dbg //plaidml/bridge/keras:backend_test -- TestBackendOps.res3a_branch2a

### CODE REVIEW

File: pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc [hunk 1]
```diff
@@ -31,13 +32,31 @@ struct StdxSubgroupBroadcastOpConversion
...
+  using SPIRVOpLowering<AllocOp>::SPIRVOpLowering;
+
+  AllocOpPattern(MLIRContext *context, SPIRVTypeConverter &typeConverter)
+      : SPIRVOpLowering<AllocOp>(context, typeConverter, /*benefit=*/1000) {}
+
+  LogicalResult
+  matchAndRewrite(AllocOp operation, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    MemRefType allocType = operation.getType();
+    MemRefType allocType2 = MemRefType::Builder(allocType).setMemorySpace(6);
```
Comment (adstraw) [inline pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc:52]: Magic number 6?

File: pmlc/dialect/pxa/transforms/gpu_thread.cc [hunk 2]
```diff
@@ -67,12 +67,22 @@ struct GPUThreadPass : public GPUThreadBase<GPUThreadPass> {
...
+      }
+      CostModel model(op, goalThreads);
       auto tileSize =
           findBestTileSize(EvenTilingGenerator(), model, *maybeRanges);
       auto inner = performTiling(op, tileSize);
       op.setAttr("hardware", StringAttr::get("gpu_block", &getContext()));
       inner.setAttr("hardware", StringAttr::get("gpu_thread", &getContext()));
+      if (subgroupAttr) {
+        op.removeAttr("subgroupSize");
+        inner.setAttr("subgroupSize", subgroupAttr);
```
Comment (adstraw) [inline pmlc/dialect/pxa/transforms/gpu_thread.cc:84]: This seems superfluous as we got subgroupAttr from the op attribute "subgroupSize", did not modify it, and now resetting it.

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 3]
```diff
@@ -230,7 +232,9 @@ struct SubgroupsPass : public SubgroupsBase<SubgroupsPass> {
         6,       // Maximum register per thread
     };
     SubgroupCostModel cm(params, op);
-    if (cm.bestCost == std::numeric_limits<double>::infinity()) {
+    if (cm.bestCost == std::numeric_limits<double>::infinity() ||
+        op.getIVs().size() < 6) {
```
Comment (adstraw) [inline pmlc/dialect/pxa/transforms/subgroups.cc:236]: Magic number 6, again.

**Review by adstraw — COMMENTED** [review]
looking good so far, just a few minor comments

File: pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc (hunk 1 above)
Comment (jbruestle) [reply pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc:52]: Yeah, this is just lack of cleanup, need to pull in the appropriate enum (6 is the function local spirv memory space).

File: pmlc/dialect/pxa/transforms/gpu_thread.cc (hunk 2 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/transforms/gpu_thread.cc:84]: We tile the block (i.e. op changes to two block, op = outer, innner = inner).  We are 'moving' the attribute off of the outer block and placing it on the inner block.  In general, I feel like this use of attributes to squirrel information from one pass to another is a poor design, Frank and I have been talking over other options, but haven't arrived at a solution yet.

File: pmlc/dialect/pxa/transforms/subgroups.cc (hunk 3 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/transforms/subgroups.cc:236]: This is a pure hack to make sure we only 'subgroup' the convolution and not the eltwise stuff so I only have one sub-grouped thing to test and will definitely not make it into the non WIP commit.

**Comment by jbruestle** [comment]
As of the latest (even hackier) changes, we get correct results for some of the resnet convolutions.  Remaining issue seems to be subgroup size mismatch on the hardware side.

**Comment by jbruestle** [comment]
Code is now tidy enough to land (hopefully).  Also, edsl test (except winograd) and many backend tests involving convolutions now work with subgroups, but due to the fact that 1) The CI machines dont have the latest driver version which is now required and 2) Resnet still doesn't quite work for subgroups, I've disabled it pmlc/target/intel_gen/pipeline.cc and replaced it with tileAccumulate for the official PR, but if you swap which pass is commented, subgroups work.

File: pmlc/dialect/pxa/transforms/gpu_thread.cc [hunk 4]
```diff
@@ -61,18 +64,26 @@ struct GPUThreadPass : public GPUThreadBase<GPUThreadPass> {
...
+    // Nest outermost loops into 'blocks' and 'threads'
     for (auto op : func.getOps<AffineParallelOp>()) {
       auto maybeRanges = op.getConstantRanges();
       if (!maybeRanges) {
+        // Fail if we can't compute the ranges at compile time
         continue;
       }
-      CostModel model(op, maxThreads);
+      // We want 'logical threads' * 'threads per subgroup' (i.e. subgroupSize)
+      // to end up beging about 'maxThreads' threads, and the rest we put into
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/gpu_thread.cc:75]: typo: 'beging'?

**Review by flaub — APPROVED** [review]
LGTM

**Pull request merged**